### PR TITLE
[pt1][quant] Quantized FC operator

### PR DESCRIPTION
**Pull request by jianyuh**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19750 [pt1][quant] Quantized FCRelu operator&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15079097/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19541 [pt1][quant] Use QTensor with quantized FC operator&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14900407/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#19497 [pt1][quant] Quantized FC operator**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14761865/)

Implements a basic quantized FC (uint8 * int8 -> uint8) with FBGEMM APIs.

Related document:
https://fb.quip.com/rP5tAx56ApMM
https://fb.quip.com/MU7aAbzGDesu

Work Item List:
1. [DONE] currently we use prepack routines inside Quantized FC operator. Will separate it as a standalone operator soon.
2. [DONE] rebase to D14817809 and D14994781 (cpp custom types).
3. [DONE] correctness unit test.

4. [To Do] rebase to QTensor. Similar to D14565413, this will be implemented in the next Diff.

Differential Revision: [D14761865](https://our.internmc.facebook.com/intern/diff/D14761865/)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/qfc.cpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+namespace at {
+namespace native {
+namespace {
+
+class QFCInt8 final : public c10::OperatorKernel {
+ public:
+#ifdef USE_FBGEMM
+  std::tuple<at::Tensor, double, int64_t> operator()(
+      const at::Tensor& input,
+      double input_scale,
```
Comment (jspark1105) [inline aten/src/ATen/native/quantized/cpu/qfc.cpp]: Is using double for scale convention in PT quantized operator? Otherwise, we can just use float here and we won't need the static_cast below.

File: aten/src/ATen/native/quantized/cpu/qfc.cpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      int64_t input_zero_point,
```
Comment (jspark1105) [inline aten/src/ATen/native/quantized/cpu/qfc.cpp]: A similar question here. int32_t or int should be good enough here.
nit: if we want to strictly follow the standard, int64_t should be std::int64_t

File: aten/src/ATen/native/quantized/cpu/qfc.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    // the same numerics across different machines. Therefore, we do not provide
+    // a fallback path and rather fail loudly if we cannot run FBGEMM.
+    AT_ASSERTM(
+        fbgemm::fbgemmSupportedCPU(), "Your CPU does not support FBGEMM.");
+
+    auto input_contig = input.contiguous();
+    const auto* input_ptr = input_contig.data<uint8_t>();
+
+    AT_ASSERT(input.dim() >= 2);
+    int64_t M = 1;
```
Comment (jspark1105) [inline aten/src/ATen/native/quantized/cpu/qfc.cpp]: Maybe add a comment here that M refers to M in the standard GEMM notation? Not to be confused with some operators implemented in some deep learning frameworks using M for the number of output channels.

File: aten/src/ATen/native/quantized/cpu/qfc.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+    float weight_scale_float = static_cast<float>(weight_scale);
+    int32_t weight_zero_point_int32 = static_cast<int32_t>(weight_zero_point);
+
+    float output_scale_float = static_cast<float>(output_scale) /
+        input_scale_float / weight_scale_float;
+    int32_t output_zero_point_int32 = static_cast<int32_t>(output_zero_point);
+
+    // This operation does the following:
+    // 1) Quantizes the input matrix given the statistics we've calculated above
```
Comment (jspark1105) [inline aten/src/ATen/native/quantized/cpu/qfc.cpp]: We're not quantizing input inside the operator and rather assumes input is already quantized no?

File: aten/src/ATen/fbgemm_utils.h [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#ifdef USE_FBGEMM
+#include "fbgemm/Fbgemm.h"
+#include "fbgemm/QuantUtils.h"
+
+// The struct for the packed weight matrix (PackBMatrix) and the corresponding
+// column offsets used for the fully connect layer, which are both prepared in
+// the prepacking step to save the computations in the inference. Note the
+// column offsets includes the sum of the B columns as well as the scalar term
```
Comment (jspark1105) [inline aten/src/ATen/fbgemm_utils.h]: Maybe worth mentioning column offset is needed for asymmetric quantization of input (affine quantization in PT).

File: aten/src/ATen/native/quantized/cpu/qfc.cpp [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    // 2) Creates a "row buffer" vector with offset values that must be added
```
Comment (jspark1105) [inline aten/src/ATen/native/quantized/cpu/qfc.cpp]: Maybe worth mentioning row buffer is needed when we use affine quantization for weights.

File: aten/src/ATen/native/quantized/cpu/qfc.cpp [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    //
+    //  Note this is not executed eagerly, but rather within the fbgemmPacked
+    //  call below.
+    fbgemm::PackAWithRowOffset<uint8_t> packA(
+        /*trans=*/fbgemm::matrix_op_t::NoTranspose,
+        /*nRow=*/M,
+        /*nCol=*/K,
+        /*smat=*/input_ptr,
+        /*ld=*/K,
+        /*pmat=*/nullptr); // packA manages ownership of `pmat`
```
Comment (jspark1105) [inline aten/src/ATen/native/quantized/cpu/qfc.cpp]: Maybe add a TODO to consider a way to pre-allocate (and reuse) pmat buffer?

File: aten/src/ATen/fbgemm_utils.h [hunk 8]
```diff
@@ -0,0 +1,21 @@
+#pragma once
```
Comment (raghuramank100) [inline aten/src/ATen/fbgemm_utils.h]: Should this file be under quantized/cpu?

File: aten/src/ATen/native/quantized/cpu/qfc.cpp [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+      int64_t output_zero_point) {
+    // uint8 * int8 -> uint8 (no quantization/dequantization)
+
+    // We make a strong guarantee that models using these operators will have
+    // the same numerics across different machines. Therefore, we do not provide
+    // a fallback path and rather fail loudly if we cannot run FBGEMM.
+    AT_ASSERTM(
+        fbgemm::fbgemmSupportedCPU(), "Your CPU does not support FBGEMM.");
+
+    auto input_contig = input.contiguous();
```
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/cpu/qfc.cpp]: Please insert a TODO at all locations where .contiguous is called for further jit optimizations

File: test/test_quantized.py [hunk 10]
```diff
@@ -4,20 +4,83 @@
...
+    )
+
+    Y_q_ref = (
+        (np.round((Prod_XqWq_ref + b_q) * (Y_scale / X_scale / W_scale)) + Y_zp)
+        .clip(0, 255)
+        .astype(np.uint8)
+    )
+    return Y_q_ref
+
+
```
Comment (raghuramank100) [inline test/test_quantized.py]: What is the purpose of the skip if not registered decorator here?

File: test/test_quantized.py [hunk 11]
```diff
@@ -4,20 +4,83 @@
...
 @skipIfNotRegistered("Relu_ENGINE_DNNLOWP",
                      "fbgemm-based Caffe2 ops are not linked")
 class TestQuantized(TestCase):
     def test_relu(self):
         a = (torch.tensor([4, 6, 1, 10], dtype=torch.uint8), 0.01, 5)
         r = torch.ops.c10.quantized_relu(a)
-        np.testing.assert_equal(r[0].numpy(), torch.tensor([5, 6, 5, 10], dtype=torch.uint8).numpy())
+        np.testing.assert_equal(
```
Comment (raghuramank100) [inline test/test_quantized.py]: Why are these changes part of this PR?

File: aten/src/ATen/native/quantized/cpu/qfc.cpp (hunk 1 above)
Comment (jspark1105) [reply aten/src/ATen/native/quantized/cpu/qfc.cpp]: Sorry, just saw that in the follow-up diff that makes my comment irrelevant. Please ignore.

File: aten/src/ATen/native/quantized/cpu/qfc.cpp (hunk 2 above)
Comment (jspark1105) [reply aten/src/ATen/native/quantized/cpu/qfc.cpp]: Same here. Please ignore.

File: test/test_quantized.py [hunk 12]
```diff
@@ -4,20 +4,83 @@
...
+from common_utils import TEST_WITH_UBSAN, TestCase, run_tests, skipIfNotRegistered
 
 
 def canonical(graph):
     return str(torch._C._jit_pass_canonicalize(graph))
 
 
+def _quantize(x, scale, zero_point):
+    """Quantizes a numpy array."""
+    qx = np.round(x / scale).astype(np.uint8) + zero_point
```
[the remark this answers is not in the document]
Comment (jspark1105) [reply test/test_quantized.py]: NVM. Saw PR 19319 was fixing this.

File: test/test_quantized.py (hunk 10 above)
Comment (jianyuh) [reply test/test_quantized.py]: raghuramank100: I think here the decorator means that if the fbgemm-based Caffe2 ops are not linked, then this unit test will be skipped. There should be some dependencies. @zafartahirov Correct me if I am wrong.

File: aten/src/ATen/native/quantized/cpu/qfc.cpp (hunk 4 above)
Comment (jianyuh) [reply aten/src/ATen/native/quantized/cpu/qfc.cpp]: Will remove the first point.

File: aten/src/ATen/fbgemm_utils.h (hunk 5 above)
Comment (jianyuh) [reply aten/src/ATen/fbgemm_utils.h]: Will add these comments.

File: aten/src/ATen/native/quantized/cpu/qfc.cpp (hunk 6 above)
Comment (jianyuh) [reply aten/src/ATen/native/quantized/cpu/qfc.cpp]: Will mention this in the comment.

File: aten/src/ATen/native/quantized/cpu/qfc.cpp (hunk 7 above)
Comment (jianyuh) [reply aten/src/ATen/native/quantized/cpu/qfc.cpp]: Will add this TODO.

File: aten/src/ATen/fbgemm_utils.h (hunk 8 above)
Comment (jianyuh) [reply aten/src/ATen/fbgemm_utils.h]: This "fbgemm_utils.h" might be reused by some routines in other places, so it is put in the same directory as "cpp_custom_type_hack.h". In addition, we need to change the TARGETS and CMakefile in order to enable the header files in "quantized/cpu" location.

File: aten/src/ATen/native/quantized/cpu/qfc.cpp (hunk 9 above)
Comment (jianyuh) [reply aten/src/ATen/native/quantized/cpu/qfc.cpp]: Will add a TODO.

File: test/test_quantized.py (hunk 11 above)
Comment (jianyuh) [reply test/test_quantized.py]: Sorry this is formatted by Nuclide. Only for the formatting reason.

File: test/test_quantized.py [hunk 13]
```diff
@@ -4,20 +4,83 @@
...
+    return qx
+
+
+def _dequantize(qx, scale, zero_point):
+    """Dequantizes a numpy array."""
+    x = (qx - zero_point) * scale
+    return x
+
+
+# Make sure we won't have overflows from vpmaddubsw instruction used in fbgemm)
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply test/test_quantized.py]: Will fix this.

File: test/test_quantized.py [hunk 14]
```diff
@@ -4,20 +4,83 @@
...
+
+    Y_q_ref = (
+        (np.round((Prod_XqWq_ref + b_q) * (Y_scale / X_scale / W_scale)) + Y_zp)
+        .clip(0, 255)
+        .astype(np.uint8)
+    )
+    return Y_q_ref
+
+
 @skipIfNotRegistered("Relu_ENGINE_DNNLOWP",
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply test/test_quantized.py]: Will change it in the next version.

File: aten/src/ATen/native/quantized/cpu/qfc.cpp (hunk 9 above)
Comment (raghuramank100) [reply aten/src/ATen/native/quantized/cpu/qfc.cpp]: Minor comment: please rephrase to: TODO: .contiguous call, needs to be optimized with JIT passes.

File: test/test_quantized.py [hunk 15]
```diff
@@ -121,5 +171,81 @@ def test_qsumrelu_different_qparams(self):
...
+        W_value_max = 127
+        W_q = np.round(
+            np.random.rand(output_channels, input_channels)
+            * (W_value_max - W_value_min)
+            + W_value_min
+        ).astype(np.int8)
+
+        b_q = np.round(np.random.randn(output_channels) * 10 - 10).astype(np.int32)
+
+        # Make sure X_scale * W_scale * input_channels * X_value_max * W_value_max
```
Comment (raghuramank100) [inline test/test_quantized.py]: Why is this condition needed? Shouldn't the test pass for any value of Y_scale?

File: test/test_quantized.py [hunk 16]
```diff
@@ -24,13 +24,58 @@ def _dequantize(qx, scale, zero_point):
     return x
 
 
-@skipIfNotRegistered("Relu_ENGINE_DNNLOWP",
+# Make sure we won't have overflows from vpmaddubsw instruction used in FBGEMM
```
Comment (raghuramank100) [inline test/test_quantized.py]: Please add some comments explaining what is going on inside this function. It looks like this function is modifying the weights to eliminate overflow, can you elaborate a bit?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please see comments below, thanks

File: test/test_quantized.py [hunk 17]
```diff
@@ -121,5 +171,81 @@ def test_qsumrelu_different_qparams(self):
...
+            X_q,
+            X_value_min,
+            X_value_max,
+            W_q,
+            W_value_min,
+            W_value_max,
+        )
+
+        # Reference quantized FC operator
+        Y_q_ref = qfc_ref(X_q, X_scale, X_zp, W_q, W_scale, W_zp, b_q, Y_scale, Y_zp)
```
Comment (raghuramank100) [inline test/test_quantized.py]: Will be great to have a test that checks a floating point fc reference operator with quantized weights, activations and biases and ensure that the output matches. Test would be something like:
W_f = W_q.dequantize()
X_f = X_q.dequantize()
Y_f = F.Linear(W_f,X_f,B_f)
assert(Y_f==Y_q_ref.dequantize())

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please consider adding an additional test to verify that quantized operators match floating point operators, provided inputs are quantized.

File: test/test_quantized.py (hunk 15 above)
Comment (jianyuh) [reply test/test_quantized.py]: This is not a necessary condition. As you mentioned, the test should pass for any values of Y_scale. However, when Y_scale is too small, we will get a result matrix Y_q/Y_q_ref with almost all entries 255. Then the assert_equal comparisons in the unit test might not make too much sense.

File: test/test_quantized.py (hunk 16 above)
Comment (jianyuh) [reply test/test_quantized.py]: Will elaborate the function.

File: test/test_quantized.py (hunk 17 above)
Comment (jianyuh) [reply test/test_quantized.py]: Due to the clamp function in the requantization from int32_t to uint_t (out[i] = clamp(in[i] * scale + zero_point, 0, 255)), there should be accuracy loss. Then Y_f should not be equal to Y_q_ref.dequantize(). Correct me if I am wrong.

**Comment by jianyuh** [comment]
@raghuramank100: As we discussed, due to the clamp function in the requantization from int32_t to uint_t (out[i] = clamp(in[i] * scale + zero_point, 0, 255)), there should be accuracy loss. We can only compare the _quantize(Y_fp32_ref, Y_scale, Y_zp) with Y_q, instead of comparing the  floating point results directly.

File: test/test_quantized.py (hunk 15 above)
Comment (jianyuh) [reply test/test_quantized.py]: Here are the results for different settings of Y_scale. Finally we choose Y_scale = 125.1234 for better test coverage. We should use python hypothesis for better testing in the future.

1. Y_scale = 125.1234

Y_q:
[[100  87   0 255  24   0 115 158]
 [108 107   0 195   0   0  38 255]
 [119  38   0 100  16 144  48 201]
 [ 94 121   0 255   0   0  84 255]]
Y_q_ref:
[[100  87   0 255  24   0 115 158]
 [108 107   0 195   0   0  38 255]
 [119  38   0 100  16 144  48 201]
 [ 94 121   0 255   0   0  84 255]]
Y_q_ref2:
[[100  87   0 255  24   0 115 158]
 [108 107   0 195   0   0  38 255]
 [119  38   0 100  16 144  48 201]
 [ 94 121   0 255   0   0  84 255]]


2. Y_scale = 1345.1234

Y_q:
[[ 0 16  0 12  1  1  0 21]
 [ 0 13  0  0  0  4  0 31]
 [ 0 17  0  6 10 15  0  0]
 [ 0  9  0 28  7  5  0 12]]
Y_q_ref:
[[ 0 16  0 12  1  1  0 21]
 [ 0 13  0  0  0  4  0 31]
 [ 0 17  0  6 10 15  0  0]
 [ 0  9  0 28  7  5  0 12]]
Y_q_ref2:
[[ 0 16  0 12  1  1  0 21]
 [ 0 13  0  0  0  4  0 31]
 [ 0 17  0  6 10 15  0  0]
 [ 0  9  0 28  7  5  0 12]]


3. Y_scale = 13.1234

Y_q:
[[255 255 252 255   0   0   0   0]
 [118 255   0 229   0   0  36   0]
 [255 255 255 111   0   0   0   0]
 [255 255   0 135   0   0   0   0]]
Y_q_ref:
[[255 255 252 255   0   0   0   0]
 [118 255   0 229   0   0  36   0]
 [255 255 255 111   0   0   0   0]
 [255 255   0 135   0   0   0   0]]
Y_q_ref2:
[[255 255 252 255   0   0   0   0]
 [118 255   0 229   0   0  36   0]
 [255 255 255 111   0   0   0   0]
 [255 255   0 135   0   0   0   0]]

**Review by raghuramank100 — APPROVED** [review]
